core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -86,20 +90,56 @@ public TableScan appendsAfter(long fromSnapshotId) {
   public TableScan useSnapshot(long scanSnapshotId) {
     Preconditions.checkArgument(
         snapshotId() == null, "Cannot override snapshot, already set to id=%s", snapshotId());
+    Preconditions.checkArgument(
+        ref() == null, "Cannot override snapshot, already set to id=%s", ref());
     Preconditions.checkArgument(
         tableOps().current().snapshot(scanSnapshotId) != null,
         "Cannot find snapshot with ID %s",
         scanSnapshotId);
     return newRefinedScan(
-        tableOps(), table(), tableSchema(), context().useSnapshotId(scanSnapshotId));
+        tableOps(),
+        table(),
+        tableSchema(),
+        context().useRef(SnapshotRef.MAIN_BRANCH).useSnapshotId(scanSnapshotId));
+  }
+
+  @Override
+  public TableScan useRef(String name) {
```

Comment:
Because we can't combine this we `asOfTime` (see my other comment), I think the implementation can be simpler. The context doesn't need to know the `ref`. This can act just like `asOfTime` and set a specific snapshot that his the head of the branch or tag.

Reply:
Updated